SwrveCommonSDK/src/main/java/com/swrve/sdk/SwrveImp.java

```diff
@@ -550,15 +553,32 @@ protected ITelephonyManager getTelephonyManager(Context context) {
         return new AndroidTelephonyManagerWrapper(context);
     }
 
-    protected void findCacheFolder(Context context) {
+    @RequiresPermission (Manifest.permission.WRITE_EXTERNAL_STORAGE)
+    protected void findCacheFolder(Activity activity) {
         cacheDir = config.getCacheDir();
+
         if (cacheDir == null) {
-            cacheDir = context.getCacheDir();
-        }
+            cacheDir = activity.getCacheDir();
+        } else {
+            if (!checkPermissionGranted(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
+                final String[] permissions = new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE};
+                requestPermissions(activity, permissions);
+                cacheDir = activity.getCacheDir(); // fall back to internal cache until permission granted.
+            }
 
-        if (!cacheDir.exists()) {
-            cacheDir.mkdirs();
+            if (!cacheDir.exists()) {
+                cacheDir.mkdirs();
+            }
         }
+        Log.d(LOG_TAG, "Using cache directory at " + cacheDir.getPath());
```

Comment:
This will need to change to SwrveLogger later. I'll make a note of it when we port these changes to 4.1